### DISCUSSION TITLE
Broadcasting shards of a tensor one shard at the time

### DISCUSSION THREAD

**Discussion by jpilaul** [jax-ml/jax#21843] [Q&A]
I am trying to broadcast one shard of `A` at the time to all devices. In the code below, `B` and `C` are sharded and put on devices the way that I want. Meaning each device has one of the shards for `B` and `C` (i.e. device 0 will get shard_0 of `B` and `C`, device 1 shard_1 of `B` and `C`, device 2 shard_2, ...). However, each device sees all of `A` when I would like all devices to first get shard_0, then on the next iteration shard_1, and so on. How can we just send shards of A sequentially?

```python
import jax
import jax.numpy as jnp
import jax.random as rand
from jax.experimental.shard_map import shard_map
from jax.sharding import PartitionSpec as P
from jax.sharding import Mesh
from functools import partial
from jax import lax

A, B, C = make_data((2, 8, 16, 32))  # all tensors are of size [2, 8, 16, 32]

devices = jax.devices(backend='gpu')
n_device = len(devices)
mesh = Mesh(np.array(devices), axis_names=('i',))
in_specs = (P(None,None,None,None), P(None,'i',None,None), P(None,'i',None,None))
out_specs=P(None,None,None)

@jax.jit
@partial(shard_map, mesh=mesh, in_specs=in_specs, out_specs=out_specs, check_rep=False)
def f(a,b,c):
    return a * b + c

A_shards = jnp.split(A, n_device, axis=1)

def scan_fn(carry, a_shard):
    output = f(a_shard, B, C)
    carry = None
    return carry, output

_, outputs = lax.scan(scan_fn, None, jnp.array(A_shards))
res = jnp.concatenate(outputs, axis=1)
```

Thank you!

tagging @nshepperd and @vasushyam

**Comment by nshepperd** [comment]
I found that bcast in mpi4jax <https://mpi4jax.readthedocs.io/en/latest/api.html#bcast> is almost the thing we want, but does this at the host rather than device level - it broadcasts from a given host to all other hosts in a multi node setup.

What we'd like to know if there's any primitive in jax+xla that could do the analogous thing for xla devices. The fact that indexing into a sharded dimension of an Array within jit is lowered to an inefficient-looking all_gather followed by slice leads me to think that this doesn't exist, but in that case - why? - and also, what would be the most efficient way to do this then (besides "don't")?
